lib/elixir/lib/exception.ex

```diff
@@ -949,28 +951,161 @@ defmodule MismatchedDelimiterError do
+        line: start_line,
+        column: start_column,
         end_line: end_line,
         end_column: end_column,
         description: description,
         file: file,
         snippet: snippet
       }) do
-    snippet =
-      :elixir_errors.format_snippet(
-        {end_line, end_column},
-        file,
-        description,
-        snippet,
-        :error,
-        [],
-        nil
-      )
+    start_pos = {start_line, start_column}
+    end_pos = {end_line, end_column}
+    lines = String.split(snippet, "\n")
 
+    snippet = format_snippet(start_pos, end_pos, description, file, lines)
     format_message(file, end_line, end_column, snippet)
   end
 
+  defp format_snippet(
+         {start_line, start_column} = start_pos,
+         {end_line, end_column} = end_pos,
+         description,
+         file,
+         lines
+       )
+       when start_line < end_line do
+    max_digits = digits(end_line)
+    general_padding = max(2, max_digits) + 1
+    padding = n_spaces(general_padding)
+
+    relevant_lines =
+      if end_line - start_line < @max_lines_shown do
+        line_range(lines, start_pos, end_pos, padding, max_digits)
+      else
+        first_line = Enum.at(lines, start_line - 1)
+        last_line = Enum.at(lines, end_line - 1)
+
+        start_padding = line_padding(start_line, max_digits)
+        end_padding = line_padding(end_line, max_digits)
+
+        """
+        #{start_padding}#{start_line} │ #{first_line}
+         #{padding}│ #{unclosed_delimiter(start_column)}
+         #{padding}│ ...
+        #{end_padding}#{end_line} │ #{last_line}
+         #{padding}│ #{mismatched_closing_delimiter(end_column)}\
+        """
+      end
+
+    """
+     #{padding}#{red("error:")} #{pad_message(description, padding)}
+     #{padding}│
+    #{relevant_lines}
+     #{padding}│
+     #{padding}└─ #{Path.relative_to_cwd(file)}:#{end_line}:#{end_column}\
+    """
+  end
+
+  defp line_padding(line_number, max_digits) do
+    line_digits = digits(line_number)
+
+    spacing =
+      if line_digits == 1 do
+        max(2, max_digits)
+      else
+        max_digits - line_digits + 1
+      end
+
+    n_spaces(spacing)
+  end
+
+  defp n_spaces(n), do: String.duplicate(" ", n)
+
+  defp digits(number, acc \\ 1)
+  defp digits(number, acc) when number < 10, do: acc
+  defp digits(number, acc), do: digits(div(number, 10), acc + 1)
+
+  defp format_snippet(
+         {start_line, start_column},
+         {end_line, end_column},
+         description,
+         file,
+         lines
+       )
+       when start_line == end_line do
+    max_digits = digits(end_line)
+    general_padding = max(2, max_digits) + 1
+    padding = n_spaces(general_padding)
+
+    line = Enum.at(lines, end_line - 1)
+    formatted_line = "#{line_padding(end_line, max_digits)}#{end_line} │ #{line}"
+
+    mismatched_closing_line = n_spaces(start_column - 1) <> red("│") <> mismatched_closing_delimiter(abs(end_column - start_column))
```

Comment:
Attempt to rely less on string concatenation and instead build lists which we will interpolate.